### ISSUE TITLE
MSYS build is broken (though not MSYS2)

### ISSUE THREAD

**Issue by pps83** [simdjson/simdjson#1066]
When I tried to compile latest simdjson with msys I got these errors:

![image](https://user-images.githubusercontent.com/1614246/88450190-f5930980-ce6e-11ea-98a8-d9aec48ec8ff.png)

After I did a quick review, some comments:

1) There is no need to check for `__MINGW64__`. `__MINGW32__` is always defined (similar to `_WIN32`)
```
#elif defined(__MINGW32__) || defined(__MINGW64__)
```

2) `SIMDJSON_VISUAL_STUDIO` check looks very confusing. If it's 1-to-1 `_MSC_VER` then it has to be `_MSC_VER` which is widely used and familiar to most devs. I actually though it was a check for `_MSC_VER` that excludes clang and other compilers that also define `_MSC_VER` on windows  If I compile with clang from cmd line on windows I'll get `SIMDJSON_VISUAL_STUDIO` even if I don't have VS installed? Also, the `posix_memalign` block makes incorrect assumption that it's available if `_MSC_VER` and `__MINGW32__` aren't defined. Also, the comment mention there `implicit defined` warning in c++. I think it would be an error in c++.

![image](https://user-images.githubusercontent.com/1614246/88449040-92e94000-ce65-11ea-8b77-f8386b1c96d2.png)

IMO, in general, to avoid all the mess with aligned malloc it's simply easier to manually code it on top of regular malloc:

```
/** @private portable version of  posix_memalign.
alignment, must be a power of two and a multiple of sizeof(void*) */
static inline void *aligned_malloc(size_t alignment, size_t size) {
  char* p = (char*)malloc(size + (alignment - 1) + sizeof(void**));
  if (p)
  {
    char *aligned_p = p + (alignment - 1) + sizeof(void**);
    aligned_p -= (uintptr_t)aligned_p & (alignment - 1);
    *((void**)(aligned_p - sizeof(void**))) = p;
    p = aligned_p;
  }
  return p;
}

static inline void aligned_free(void *mem_block) {
  if (mem_block)
    free( *(((void **)mem_block) - 1));
}
```

The other minor thing: I find it totally useless and unusual to have `@private` comments. Shouldn't all these "private" things then be placed in some `simdjson::impl` namespace. This way, if I type `simdjson::` I won't get tons on irrelevant "private" mess.

**Comment by pps83** [comment]
In general MSYS build isn't terribly important, normally MINGW should be used instead. I simply used it to run `amalgamate.sh` and quickly tries to check that all is good with the cmd line that amalgamate.sh provides. Also, note, the amalgamate.sh still has that annoying bug with line endings: it produces tons of spurious changes if anya/all files checked out with windows line endings.

**Comment by pps83** [comment]
![image](https://user-images.githubusercontent.com/1614246/88453097-ea97a380-ce85-11ea-9edd-1e8ee185f4d9.png)

missing

```
#include <strings.h>
```

**Comment by pps83** [comment]
> ... note, the amalgamate.sh still has that annoying bug with line endings: it produces tons of spurious changes if any/all files checked out with windows line endings.

I propose to force Unix line endings for c/c++ code, this way `amalgamate.sh` produces the same results on windows: https://github.com/simdjson/simdjson/pull/1069

**Comment by lemire** [comment]
I don't know anything about MSYS... it seems that there a web site that hasn't been updated in years. We don't have it in CI. 

For us to support it, we should have it in CI.


Yet it seems that the MSYS2 built broke some time ago. I don't understand. It should not break, we have CI tests.

**Comment by lemire** [comment]
@pps83 

>  If I compile with clang from cmd line on windows I'll get SIMDJSON_VISUAL_STUDIO even if I don't have VS installed?

How can you compile with clang under Windows without Visual Studio? You can certainly compile *outside* of the Visual Studio GUI, but that's true of cl as well. But you still need Visual Studio (the tools).

**Comment by TkTech** [comment]
@pps83 Are you attempting to use clang-cl or clang? Clang-cl intentionally pretends to be MSVC, and implements many of the same quirks and defines you'd expect from MSVC.

**Comment by lemire** [comment]
As @TkTech points out... what we support is clancl which is a hybrid of Visual Studio and pure LLVM clang. It is an important target for us. It is  Visual Studio *and* clang.


I am not aware that you can compile using the pure LLVM clang under windows... excluding WSL and MSYS/cygwin of course.

**Comment by lemire** [comment]
Our CI tests pass under MSYS2...
https://github.com/simdjson/simdjson/pull/1073/checks?check_run_id=910256858

I want to stress the importance of having good CI platform coverage. If MSYS is not in CI, we won't be able to support it. Does any CI system supports MSYS? We have MSYS2...

**Comment by lemire** [comment]
I removed the bug tag because we do not claim to support MSYS. I have never even used it. So it is not a bug. 

I would be happy to add MSYS as a supported platform.

**Comment by lemire** [comment]
This should not have been closed. It remains open.

**Comment by pps83** [comment]
> Our CI tests pass under MSYS2...
> https://github.com/simdjson/simdjson/pull/1073/checks?check_run_id=910256858
> 
> I want to stress the importance of having good CI platform coverage. If MSYS is not in CI, we won't be able to support it. Does any CI system supports MSYS? We have MSYS2...

Msys is read, I do use msys2 off course. `__MSYS__` is the define enabled by msys2 when building for msys2. Msys2 has to separate targets: mingw and msys, mingw is pure windows api target, msys is somewhat similar to cygwin and allows easier porting of unix utils for windows.

**Comment by lemire** [comment]
@pps83 Can you have a look at our CI tests? Look at github/workflows. We have several msys2 tests.

**Comment by pps83** [comment]
If you type [winkey] `MSYS` you'll see shortcuts for different environements:

![image](https://user-images.githubusercontent.com/1614246/88483182-03887d80-cf88-11ea-9924-dcd59e7d86f9.png)

mingw32/mingw64 work ok, but msys fails

**Comment by pps83** [comment]
> Look at github/workflows

How do I look there? Can you please link it?

**Comment by lemire** [comment]
> Maybe try to add "MSYS" there?

Pull request invited.

**Comment by pps83** [comment]
> > Maybe try to add "MSYS" there?
> 
> Pull request invited.

https://github.com/simdjson/simdjson/pull/1075 Not sure if this will work, should fail MSYS build now

**Comment by lemire** [comment]
As best as I can understand, you are using MSYS2 to build software for MSYS, a dead system. Is that it?

Can you motivate this issue... I do not understand. You stated yourself that MSYS was dead. Why would we spend time and effort supporting it?

Do you need support for it? Who needs support for MSYS?

There is a cost to supporting a platform. I am all for supporting even niche platforms like freebsd, but not dead ones.

**Comment by lemire** [comment]
Ok. So I think MSYS is a legacy 32-bit system. If that's the case, if I am correct, then we will *not* support it.

We selectively, and secretly, support some 32-bit system. But we are not going to support a marginal 32-bit system.

Please motivate the issue.

**Comment by pps83** [comment]
> As best as I can understand, you are using MSYS2 to build software for MSYS, a dead system. Is that it?

No, MSYS is basically MSYS2. All the tools there are build for MSYS which provides some minimal level of support for posix like layer to be able to port unix tools (like bash etc). Everything in msys2 is build for MSYS subsystem, everything in `C:\msys64\usr\bin` links to `msys-2.0.dll` that provides this MSYS subsystem, similar to cygwin1.dll for CYGWIN subsystem in cygwin. msys2 is the project name, they simply use the same subsystem name as the old original msys. old original msys is dead, this entire MSYS thing isn't related to the old msys in any way.

In short, imagine there is some unix app and one wants to port it to msys2, then they would need to build it for MSYS. If the app uses simdjson, then it will fail to build. The app cannot be build with mingw/mingw64 to be usable under msys2, as it won't be able to understand unix-like paths used under msys2

**Comment by lemire** [comment]
In CI, we know test with these configurations...

```
        mingw: ["MINGW32", "MINGW64", "MSYS"]
```

So we are good then and we can close this issue, right?

**Comment by lemire** [comment]
For people who are interested in MSYS2 as a platform, the documented approach to build applications can be found on our wiki at:


https://github.com/simdjson/simdjson/wiki/MSYS2

**Comment by pps83** [comment]
> In CI, we know test with these configurations...
> 
> ```
>         mingw: ["MINGW32", "MINGW64", "MSYS"]
> ```
> 
> So we are good then and we can close this issue, right?

MSYS build still fails with latest simdjson:

![image](https://user-images.githubusercontent.com/1614246/88499423-96a4d000-cfe7-11ea-8c5b-f73efa4c1501.png)

Supposedly by adding "MSYS" the test should have failed, no idea where and how to check if it actually tries to build for msys.
